lib/src/modules/gates.dart

```diff
@@ -513,3 +513,64 @@ class Mux extends Module with InlineSystemVerilog {
     return '$control ? $d1 : $d0';
   }
 }
+
+/// A generic two-input bit index gate [Module].
+///
+/// It always takes two inputs and has one output of width 1.
+class IndexGate extends Module with InlineSystemVerilog {
+  late final String _a, _b, _y;
+
+  /// The primary input to this gate.
+  Logic get a => input(_a);
+
+  /// The bit index for this gate.
+  Logic get b => input(_b);
+
+  /// The output of this gate.
+  Logic get y => output(_y);
+
+  /// Constructs a two-input bit index gate for an abitrary custom functional implementation.
+  ///
+  /// The bit [_b] will be indexed as an output.
```

Comment:
The variable `_b` is private (since it starts with `_`) so shouldn't be referenced in the doc comment for a public constructor.  This sentence is a little confusing also, could be clarified like "[selected] is a one-bit [Logic] equal to the [index]th bit of [original]".  The first sentence about "arbitrary" doesn't make sense for this module.